File: Character.java
package YushanoBoken.Model;
import java.util.Random;

public abstract class Character implements Attack, RecieveDamage, isDead {
    protected String name;
    protected int hp;
    protected int attack;
    protected int defense;
    protected int level;
    protected int exp;
    private boolean isDisable;
    private int damageAfterDefense;
    private StatusEffect effect;
    private int durationOfEffect = 1;
    private String effectDisplay;

    public int Attack(){
        return attack;
    }

    public void recieveDamage(int damage){
        if(damage <= 0)return;
        int defenseFactor = 2;
        int reduceDefense = this.defense/defenseFactor;
        int damageAfterDefense = Math.max(1, damage - reduceDefense);
        this.damageAfterDefense = damageAfterDefense;
        setHp(Math.max(0, this.hp - damageAfterDefense));
    }

    public boolean isDead(){
        return hp <= 0;
    }

    public void setEffect(StatusEffect effect){
        this.effect = effect;
        this.durationOfEffect = 1;
        switch(effect){
            case Burn:
                System.out.println(name + " has been burned. It will take damage for 3 turns.");
                this.effectDisplay = "\uD83D\uDD25";
                break;
            case Freeze:
                System.out.println(name + " has been frozen. It will not be able to move for 2 turns.");
                this.effectDisplay = "❄\uFE0F";
                isDisable = true;
                break;
            case Paralysis:
                System.out.println(name + " has been paralyzed. It may lose turns for 3 turns.");
                this.effectDisplay = "⚡";
                break;
            case Shield:
                System.out.println(name + " has activated a shield. He will take half damage for 5 turns.");
                this.effectDisplay = "\uD83C\uDF2A\uFE0F";
                break;
            case Stop:
                System.out.println(name + " has been stopped. You will lose 3 turns in a row.");
                this.effectDisplay = "⌛";
                isDisable = true;
                break;
            case Heal:
                setHp(this.hp + 8);
                System.out.println(name + " has been healed for 8 health points.");
                break;
            case RADIATION:
                System.out.println(name + " has been radiated and gonna lose 8 health points.");
                break;
        }
    }

    public void applyStatusEffect() {
        if (effect == null) return;
        switch (effect) {
            case Burn:
                if (durationOfEffect <= 3) {
                    setHp(this.hp - 1);
                }else {
                    this.effectDisplay = null;
                }
                break;
            case Freeze:
                if (durationOfEffect <= 2) {
                    isDisable = true;
                    System.out.println(name + " is frozen and cannot move.");
                } else {
                    isDisable = false;
                    System.out.println(name + " has thawed.");
                    this.effectDisplay = null;
                }
                break;
            case Paralysis:
                if (durationOfEffect <= 3) {
                    Random rand = new Random();
                    if (rand.nextInt(100) < 50) {
                        isDisable = true;
                        System.out.println(name + " is paralyzed and cannot move.");
                    } else {
                        isDisable = false;
                        this.effectDisplay = null;
                    }
                }
                break;
            case Shield:
                if (durationOfEffect <= 5) {
                    System.out.println(name + " is protected. Takes half damage.");
                } else {
                    System.out.println(name + " is no longer protected.");
                    effect = null;
                    this.effectDisplay = null;
                }
                break;
            case Stop:
                if (durationOfEffect <= 3) {
                    isDisable = true;
                    System.out.println(name + " is affected by Stop and loses this turn.");
                } else {
                    isDisable = false;
                    System.out.println(name + " has been released from the Stop effect.");
                    effect = null;
                    this.effectDisplay = null;
                }
                break;
            case RADIATION:
                if (durationOfEffect <= 3) {
                    setHp(this.hp - 8);
                }else {
                    this.effectDisplay = null;
                }
                break;
        }
        durationOfEffect++;
        if (effect != StatusEffect.Shield && effect != StatusEffect.Stop && durationOfEffect > 3) {
            effect = null;
            System.out.println(name + " no longer has any negative effects");
        }
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getHp() {
        return hp;
    }

    public void setHp(int hp) {
        this.hp = hp;
    }

    public int getAttack() {
        return attack;
    }

    public void setAttack(int attack) {
        this.attack = attack;
    }

    public int getDefense() {
        return defense;
    }

    public void setDefense(int defense) {
        this.defense = defense;
    }

    public int getLevel() {
        return level;
    }

    public void setLevel(int level) {
        this.level = level;
    }

    public int getExp() {
        return exp;
    }

    public void setExp(int exp) {
        this.exp = exp;
    }

    public void setDisable(boolean disable) {
        isDisable = disable;
    }

    public boolean isDisable() {
        return isDisable;
    }

    public int getDamageAfterDefense() {
        return damageAfterDefense;
    }

    public String getEffectDisplay() {
        return effectDisplay;
    }
}

File: RPG/Character.cpp
#include "Character.h"
#include <algorithm>
#include <cstdlib>

Character::Character(const std::string& name, int maxHp, int attack, int defense,
    const char* spritePath)
    : name(name), maxHp(maxHp), hp(maxHp), attack(attack), defense(defense),
    level(1), exp(0), isDisable(false), damageAfterDefense(0),
    effect(StatusEffect::NONE), durationOfEffect(0), position{ 0, 0 },
    framesCounter(0), currentFrame(0) {

    // Load sprite
    if (spritePath) {
        sprite = LoadTexture(spritePath);
        frameRec = { 0, 0, (float)sprite.width / 4, (float)sprite.height };
    }

    // Load sound
    
}

Character::~Character() {
    if (sprite.id != 0) UnloadTexture(sprite);
    if (hurtSound.frameCount != 0) UnloadSound(hurtSound);
}

int Character::CalculateDamage() const {
    int baseDamage = attack;

    // Status effect modifiers
    if (effect == StatusEffect::BURN && durationOfEffect <= 3) {
        baseDamage = (int)(baseDamage * 0.8f); // 20% reduction
    }

    return baseDamage;
}

void Character::ReceiveDamage(int damage) {
    if (damage <= 0) return;

    // Shield reduces damage by 50%
    if (effect == StatusEffect::SHIELD && durationOfEffect <= 5) {
        damage = (int)(damage * 0.5f);
    }

    damageAfterDefense = std::max(1, damage - (defense / 2));
    hp = std::max(0, hp - damageAfterDefense);

    if (hurtSound.frameCount != 0) {
        PlaySound(hurtSound);
    }
}

bool Character::IsDead() const {
    return hp <= 0;
}

void Character::SetEffect(StatusEffect newEffect) {
    effect = newEffect;
    durationOfEffect = 1;

    switch (effect) {
    case StatusEffect::BURN:
        effectDisplay = u8"🔥";
        TraceLog(LOG_INFO, "%s has been burned!", name.c_str());
        break;

    case StatusEffect::FREEZE:
        effectDisplay = u8"❄️";
        isDisable = true;
        TraceLog(LOG_INFO, "%s has been frozen!", name.c_str());
        break;

    case StatusEffect::PARALYSIS:
        effectDisplay = u8"⚡";
        TraceLog(LOG_INFO, "%s has been paralyzed!", name.c_str());
        break;

    case StatusEffect::SHIELD:
        effectDisplay = u8"🛡️";
        TraceLog(LOG_INFO, "%s activated a shield!", name.c_str());
        break;

    case StatusEffect::STOP:
        effectDisplay = u8"⏳";
        isDisable = true;
        TraceLog(LOG_INFO, "%s has been stopped!", name.c_str());
        break;

    case StatusEffect::HEAL:
        hp = std::min(maxHp, hp + 8);
        TraceLog(LOG_INFO, "%s healed 8 HP!", name.c_str());
        break;

    case StatusEffect::RADIATION:
        effectDisplay = u8"☢️";
        TraceLog(LOG_INFO, "%s has been irradiated!", name.c_str());
        break;

    default:
        break;
    }
}

void Character::ApplyStatusEffect() {
    if (effect == StatusEffect::NONE) return;

    switch (effect) {
    case StatusEffect::BURN:
        if (durationOfEffect <= 3) {
            hp = std::max(0, hp - 1);
            TraceLog(LOG_INFO, "%s takes 1 burn damage!", name.c_str());
        }
        else {
            effectDisplay.clear();
        }
        break;

    case StatusEffect::FREEZE:
        if (durationOfEffect <= 2) {
            isDisable = true;
        }
        else {
            isDisable = false;
            effectDisplay.clear();
            TraceLog(LOG_INFO, "%s thawed out!", name.c_str());
        }
        break;

    case StatusEffect::PARALYSIS:
        if (durationOfEffect <= 3) {
            if (rand() % 100 < 50) {
                isDisable = true;
                TraceLog(LOG_INFO, "%s is paralyzed!", name.c_str());
            }
        }
        else {
            effectDisplay.clear();
        }
        break;

    case StatusEffect::RADIATION:
        if (durationOfEffect <= 3) {
            hp = std::max(0, hp - 8);
            TraceLog(LOG_INFO, "%s takes 8 radiation damage!", name.c_str());
        }
        else {
            effectDisplay.clear();
        }
        break;

    case StatusEffect::SHIELD:
        if (durationOfEffect > 5) {
            effect = StatusEffect::NONE;
            effectDisplay.clear();
            TraceLog(LOG_INFO, "%s's shield expired!", name.c_str());
        }
        break;

    case StatusEffect::STOP:
        if (durationOfEffect > 3) {
            isDisable = false;
            effect = StatusEffect::NONE;
            effectDisplay.clear();
            TraceLog(LOG_INFO, "%s can move again!", name.c_str());
        }
        break;

    default:
        break;
    }

    durationOfEffect++;
}

void Character::BasicAttack(Character& target) {
    int damage = CalculateDamage();
    target.ReceiveDamage(damage);
    TraceLog(LOG_INFO, "%s attacks %s for %d damage!",
        name.c_str(), target.GetName().c_str(), damage);
}

void Character::UpdateAnimation() {
    framesCounter++;
    if (framesCounter >= 12) { // Adjust for animation speed
        framesCounter = 0;
        currentFrame++;

        if (currentFrame > 3) currentFrame = 0;

        frameRec.x = (float)currentFrame * (float)sprite.width / 4;
    }
}

void Character::Draw() const {
    if (sprite.id != 0) {
        // Draw character sprite with current animation frame
        DrawTextureRec(sprite, frameRec, position, WHITE);

        // Draw status effect icon
        if (!effectDisplay.empty()) {
            DrawText(effectDisplay.c_str(),
                position.x + frameRec.width / 2 - 10,
                position.y - 30, 20, RED);
        }
    }
}

void Character::DrawHealthBar(Vector2 pos) const {
    float healthPercentage = (float)hp / (float)maxHp;
    Color healthColor = healthPercentage > 0.6f ? GREEN :
        healthPercentage > 0.3f ? YELLOW : RED;

    // Background
    DrawRectangle(pos.x, pos.y, 104, 14, GRAY);
    // Health bar
    DrawRectangle(pos.x + 2, pos.y + 2, (int)(100 * healthPercentage), 10, healthColor);
    // Text
    DrawText(TextFormat("%d/%d", hp, maxHp), pos.x + 30, pos.y + 2, 10, BLACK);
}
